README
                                 wgrib2           revised 7.2016

wgrib2 is a program to read/write grib2 files.


For documentation, see

    http://www.cpc.ncep.noaa.gov/products/wesley/wgrib2/index.html


For documentation of commonly used program options, type

    wgrib2


For documentation of all program options, type

   wgrib2 -help all

For documenation of any program option that is poorly described on-line
or by the "wgrib2 -help all", check the source code by searching
for "f_{option}" where the option does not include the minus sign.


Default makefile options

	USE_NETCDF3=1
	USE_NETCDF4=0
	USE_REGEX=1
	USE_TIGGE=1
	USE_MYSQL=0
	USE_IPOLATES=0
	USE_UDF=0
	USE_OPENMP=1
	USE_PROJ4=0
	USE_WMO_VALIDATION=0
	DISABLE_TIMEZONE=0
	MAKE_FTN_API=0

	USE_G2CLIB=0
	USE_PNG=1
	USE_JASPER=1
	USE_AEC=1

To compile without netcdf,    set USE_NETCDF3=0 and USE_NETCDF4=0
To compile with netcdf v3, set USE_NETCDF3=1 and USE_NETCDF4=0
To compile with netcdf v4, set USE_NETCDF3=0 and USE_NETCDF4=1

For netcdf4, the netcdf4 and hdf5 libraries are usually not included 
  in the wgrib2.tgz file to save space.  The makefile suggests
  how to download the netcdf4 and hdf5 libraries.

If you have non-POSIX C compiler system that doesn't include <regex.h>, then
you have to set USE_REGEX=0.  This disables the use of regular expressions. 

No harm to keep TIGGE support on.  USE_TIGGE=1

Support for Mysql is an option.  You have to modify the makefile to indicate 
the locations of the mysql includes and libraries and set USE_MYSQL=1

The option -new_grid uses the ipolates library to do the interpolation.
The required libraries are written in fortran and a few compilers are
already supported in the makefile.  For other compilers, you are on
your own.  The source code and makefile will have to be modified to use 
the ipolates option.  Consult a local expert if you want to install this 
optional package.  No help is available from NCEP for installing the package.

User Defined Functions (UDF) allow you to run shell commands from
within wgrib2. UDF are not available on windows machines unless
<unistd.h> is implimented by the compiler (cygwin, MinGW).  For
the security paranoid, turn off UDF by USE_UDF=0.  

OpenMP allows wgrib2 to run on many cores for a faster run time.  See 
README.openmp.  The USE_OPENMP option will compile wgrib2 with openmp 
for several compilers.

Proj4 is a open-source geolocation library.  Usually the faster gctpc library
is used for geolocation.  At the current time, Proj4 does not add any
capabilty over gctpc.

Wgrib2 has been used to test proposals to the WMO grib standard.  Developers
might want to have conditional compilation of the tests using 

        #ifdef WMO_VALIDATION
            ... test code ...
        #endif

and compile using

	USE_WMO_VALIDATION=1


If you have a non-POSIX C compiler system that doesn't support time zones
or supports time zones in a non-POSIX manner, then you have to set
DISABLE_TIMEZONE to 1.

Wgrib2 is both a stand alone utility and a library that is callable
from both C and Fortran.  To enable the making the wgrib2 library, you
have to set

	MAKE_FTN_API=1

In older versions of wgrib2, the g2clib was the default decoder of grib files.
In the current version, you can use g2clib as an optional decoder.  The main 
use of compiling wgrib2 with g2clib is for testing g2clib.   

	USE_G2CLIB=1

Support for the png, jpeg2000 and CDCSD(AEC) compression can be turned off through 
the following options.

	USE_PNG=0
	USE_JASPER=0
        USE_AEC=0


You might want to turn off the various compressions because

   1) libraries do not compile correctly (icc and pgcc have problems with Jasper)
   2) reduce the executable size and compile time
   3) problems with cross-compiling



wgrib2/Changes
pre-release  1/2004: Variable names now appear on the inventory. Ok, only 1% of the GRIB names have 
                been typed in and I plan to revise this code.
pre-release  2/2004: Levels now work. Ok some levels may need polishing but it works. Inventory of 
                 sample eta file lacks grid information but is usable.
pre-release  9/2004: Jaakko Hyvätti added the "gribtab" code
pre-release  8/2005: extraction of grib data works
v 0.1        9/2005: alpha release
v 0.1.1  12/15/2005  added scan flags for various grids, fixed some scan code,
             malloced a large array that caused a seg fault if stack was too small
v0.1.1a  4/28/2006 fix for lambert2ll Naoya Suda -  Thanks, remove some debug statements
v0.1.1b  5/4/2006 gaussian grid Sec3.c, grb2.h, rename nxny() to get_nxny() better grid output
v0.1.1c  5/23/2006 added -processid option (ex which fcst model)
         5/24/2006 added -range option for partial http transfers
v0.1.1.d 5/31/2006 change in -grid .. location of print nl
v0.1.1.e 6/2006 some time and level upgrades
v0.1.2   6/2006 -text now prints "nx ny" rather than "ndata 1", fixed binary/text output of WE:NS files
         7/2006   added the gaussian/polar/merc projections, make various FILE.c routines reentrant
         7/2006   added more packing types
	 7/2006   -flush -> flush buffers for -lola lola
	 8/2006   -del renamed to -colon, modified read_inventory and write "rec_no:location"
		  routines so that -colon , would work .. make comma seperated files
		  added info about spherical harmonics
		  moved PackVals into packing, triggered by verbose mode
		  better info about various Gaussian and lat-lon grids
		  decided to get critical values by code_table_X_Y(sec) and flag_table_X_Y(sec) calls
		  changed (projection)2ll calls to use sec rather than gdas
		  changed uint to unsigned int .. gcc extension
		  fixed -lola  bug report from D. Duncalf
		  added -no_append -no_flush, renamed -nof77 to -no_f77 to be consistent with new names
v0.1.3	 9/2006   added -vt and -VT verf time
		  added warning to bad lat-lon grid
		  better testing for lat lon values
		  mod for AIX compilation
v0.1.4   9/2006   fixed problem with decoding of fields with bitmaps
v0.1.4a		  added a test for failed memory allocation
		  change to makefile for people without . on their path (suggested by Kristian Nilssen)
	 9/2006   change for Windows geo.c allocate memory explicitly
                         closest() needs to be made faster
v0.1.4b  9/2006   fixed comment for f_order, default order is we:sn not we:ns
		  altered text_fmt to change format in init stage
                  added -verf which is a macro like -s but uses -vt rather than -t
	10/2006   f_lola, f_lat can take negative longitudes
v0.1.4c	10/2006   made f_lola much faster
	10/2006   f_lola uses append flag
	10/2006   grid type stubs,
v0.1.4d 11/2006   add order we:ns, add error check for lat-lon with -order != we:sn
		  fix some minor scan code bugs
	11/2006   text_format, text_col and header are used by -lola with text option
                  added the -is -not options
v0.1.4e	11/2006   fixed bug in un-scanning code: WE|EW:SN, (we:ns and we:sn were ok)
		  added code for complex packing with spatial diff. undefined values
		  added -match -not and -last
		  upgraded g2clib 1.0.4
		  -f77 and -no_f77 are renamed to -header and -no_header, last name change, promise
		  change to the "main" makefile g2clib-1.0.2 -> g2clib-1.0.4
	12/2006   added flag_table_3.5/3.9/3.10
	12/2006   modified -t and -vt to produce datecode in grads format when -v2 used
v0.1.4f 1/2007    Many mods from Manfred Schwarb, remove wrtieee_header, libg2c -> libgrib2c,
                  error checks for thinned grids
                  error check on verftime 
                  many cleanup patches form M. Schwarb - no effect on code operation
                  better thinned grid support .. in lat-lon grids
		  uint8 fixed for numbers > 2 GB (ok, if I had 2GB grib record, I could check the fix)
		  scan of spectral fields changed to unknown, scan handles some thinned grids
		  ilat works with thinned grids, ijlat doesn't
                  fixed Level.c () m under ground, removed factor of 1/100.  found by Caser Tejeda Hernandez
		  binary files written by -Lola now have header support
		  many patches by M Schwarb (31 for this release)!
v0.1.4gbeta 1/2007    Manfred Schwarb wrote program to download tables from the web documentation.
                  The code writes a code fragment (body of a switch statement) which is then
                  included by the code.
		    code_table0.dat    CodeTable_1.3.dat  CodeTable_3.1.dat  ncep_tableC.dat
		    CodeTable_1.2.dat  CodeTable_1.4.dat  gribtab.dat
v0.1.4g 2/2007    -help (string) is now case insensitive, -ens for ensemble info, -s includes -ens, aliases,

v0.1.4h 2/2007     more tables, process is now an alias
v0.1.5 3/2007      updated ftime so to handle product definition template 4.8 better (still has rough points)
		   replace exit() with fatal_error
		   moved #define MATCH to makefile
                   add netcdf, thanks Kristian Nilssen
		   added code table 4.10
		   added ensemble info to -verf
v0.1.5a 4/2007     Netcdf conversion to time fixed, a local-UTC conversion problem
v0.1.5b 4/2007     Netcdf module update: multiple times allowed, by Sergey Varlamov
                   cname - multiple gribtable support, -tigge for modified TIGGE names
                   Help: output format change
		   some more ensemble support: derived products, pdt=12, N_ens,
v0.1.5c 6/2007     updated Netcdf.c  by Sergey Varlamov
	7/2007     modified grib2/makefile and grib2/wgrib2/makefile (fixed clean)
        8/2007     update to makefile -l netcdf -> -lnetcdf (S. Varlamov)
                   netcdf: update format, more metadata (S. Varamov)
                   Sec3 - mercator grid had wrong variable dimension location (Boi Vuong)
                   Support for grid 204, some other minor grids added

wgrib2/Level.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "grb2.h"
#include "wgrib2.h"
#include "fnlist.h"

/* Levels.c
 *   2006: public domain wesley ebisuzaki
 *   1/2007: cleanup M. Schwarb
 *   1/2007: Caser Tejeda Hernandez found error in meter underground
 *   2/2007: level 11
 *   2/2007: spelling error fixed
 *   9/2008: type 241 added Nick Lott 
 */


/*
 * HEADER:200:lev:inv:0:level (code table 4.5)
 */

/* code table 4.5 */

const char *level_table[192] = {
/* 0 */ "reserved",
/* 1 */ "surface",
/* 2 */ "cloud base",
/* 3 */ "cloud top",
/* 4 */ "0C isotherm",
/* 5 */ "level of adiabatic condensation from sfc",
/* 6 */ "max wind",
/* 7 */ "tropopause",
/* 8 */ "top of atmosphere",
/* 9 */ "sea bottom",
/* 10 */ "entire atmosphere",
/* 11 */ "cumulonimbus base",
/* 12 */ "cumulonimbus top",
/* 13 */ "lowest level %g%% integrated cloud cover",
/* 14 */ "level of free convection",
/* 15 */ "convection condensation level",
/* 16 */ "level of neutral buoyancy",
/* 17 */ "reserved",
/* 18 */ "reserved",
/* 19 */ "reserved",
/* 20 */ "%g K level",
/* 21 */ "lowest level > %g kg/m^3",
/* 22 */ "highest level > %g kg/m^3",
/* 23 */ "lowest level > %g Bq/m^3",
/* 24 */ "highest level > %g Bg/m^3",
/* 25 */ "highest level > %g dBZ",
/* 26 */ "reserved",
/* 27 */ "reserved",
/* 28 */ "reserved",
/* 29 */ "reserved",
/* 30 */ "reserved",
/* 31 */ "reserved",
/* 32 */ "reserved",
/* 33 */ "reserved",
/* 34 */ "reserved",
/* 35 */ "reserved",
/* 36 */ "reserved",
/* 37 */ "reserved",
/* 38 */ "reserved",
/* 39 */ "reserved",
/* 40 */ "reserved",
/* 41 */ "reserved",
/* 42 */ "reserved",
/* 43 */ "reserved",
/* 44 */ "reserved",
/* 45 */ "reserved",
/* 46 */ "reserved",
/* 47 */ "reserved",
/* 48 */ "reserved",
/* 49 */ "reserved",
/* 50 */ "reserved",
/* 51 */ "reserved",
/* 52 */ "reserved",
/* 53 */ "reserved",
/* 54 */ "reserved",
/* 55 */ "reserved",
/* 56 */ "reserved",
/* 57 */ "reserved",
/* 58 */ "reserved",
/* 59 */ "reserved",
/* 60 */ "reserved",
/* 61 */ "reserved",
/* 62 */ "reserved",
/* 63 */ "reserved",
/* 64 */ "reserved",
/* 65 */ "reserved",
/* 66 */ "reserved",
/* 67 */ "reserved",
/* 68 */ "reserved",
/* 69 */ "reserved",
/* 70 */ "reserved",
/* 71 */ "reserved",
/* 72 */ "reserved",
/* 73 */ "reserved",
/* 74 */ "reserved",
/* 75 */ "reserved",
/* 76 */ "reserved",
/* 77 */ "reserved",
/* 78 */ "reserved",
/* 79 */ "reserved",
/* 80 */ "reserved",
/* 81 */ "reserved",
/* 82 */ "reserved",
/* 83 */ "reserved",
/* 84 */ "reserved",
/* 85 */ "reserved",
/* 86 */ "reserved",
/* 87 */ "reserved",
/* 88 */ "reserved",
/* 89 */ "reserved",
/* 90 */ "reserved",
/* 91 */ "reserved",
/* 92 */ "reserved",
/* 93 */ "reserved",
/* 94 */ "reserved",
/* 95 */ "reserved",
/* 96 */ "reserved",
/* 97 */ "reserved",
/* 98 */ "reserved",
/* 99 */ "reserved",
/* 100 */ "%g mb",
/* 101 */ "mean sea level",
/* 102 */ "%g m above mean sea level",
/* 103 */ "%g m above ground",
/* 104 */ "%g sigma level",
/* 105 */ "%g hybrid level",
/* 106 */ "%g m underground",
/* 107 */ "%g K isentropic level",
/* 108 */ "%g mb above ground",
/* 109 */ "PV=%g (Km^2/kg/s) surface",
/* 110 */ "reserved",
/* 111 */ "%g Eta level",
/* 112 */ "reserved",
/* 113 */ "%g logarithmic hybrid level",
/* 114 */ "snow level",
/* 115 */ "%g sigma height level",
/* 116 */ "reserved",
/* 117 */ "mixed layer depth",
/* 118 */ "%g hybrid height level",
/* 119 */ "%g hybrid pressure level",
/* 120 */ "reserved",
/* 121 */ "reserved",
/* 122 */ "reserved",
/* 123 */ "reserved",
/* 124 */ "reserved",
/* 125 */ "reserved",
/* 126 */ "reserved",
/* 127 */ "reserved",
/* 128 */ "reserved",
/* 129 */ "reserved",
/* 130 */ "reserved",
/* 131 */ "reserved",
/* 132 */ "reserved",
/* 133 */ "reserved",
/* 134 */ "reserved",
/* 135 */ "reserved",
/* 136 */ "reserved",
/* 137 */ "reserved",
/* 138 */ "reserved",
/* 139 */ "reserved",
/* 140 */ "reserved",
/* 141 */ "reserved",
/* 142 */ "reserved",
/* 143 */ "reserved",
/* 144 */ "reserved",
/* 145 */ "reserved",
/* 146 */ "reserved",
/* 147 */ "reserved",
/* 148 */ "reserved",
/* 149 */ "reserved",
/* 150 */ "%g generalized vertical height coordinate",
/* 151 */ "soil level %g",
/* 152 */ "reserved",
/* 153 */ "reserved",
/* 154 */ "reserved",
/* 155 */ "reserved",
/* 156 */ "reserved",
/* 157 */ "reserved",
/* 158 */ "reserved",
/* 159 */ "reserved",
/* 160 */ "%g m below sea level",
/* 161 */ "%g m below water surface",
/* 162 */ "lake or river bottom",
/* 163 */ "bottom of sediment layer",
/* 164 */ "bottom of thermally active sediment layer",
/* 165 */ "bottom of sediment layer penetrated by thermal wave",
/* 166 */ "maxing layer",
/* 167 */ "bottom of root zone",
/* 168 */ "ocean model level %g",
/* 169 */ "ocean level %g (kg*m-3) density difference to near surface",
/* 170 */ "ocean level %g (K) pot. temperature difference to near surface",
/* 171 */ "reserved",
/* 172 */ "reserved",
/* 173 */ "reserved",
/* 174 */ "top surface of ice on sea, lake or river",
/* 175 */ "top surface of ice, und snow on sea, lake or river",
/* 176 */ "bottom surface ice on sea, lake or river",
/* 177 */ "deep soil",
/* 178 */ "reserved",
/* 179 */ "top surface of glacier ice and inland ice",
/* 180 */ "deep inland or glacier ice",
/* 181 */ "grid tile land fraction as a model surface",
/* 182 */ "grid tile water fraction as a model surface",
/* 183 */ "grid tile ice fraction on sea, lake or river as a model surface",
/* 184 */ "grid tile glacier ice and inland ice fraction as a model surface",
/* 185 */ "reserved",
/* 186 */ "reserved",
/* 187 */ "reserved",
/* 188 */ "reserved",
/* 189 */ "reserved",
/* 190 */ "reserved",
/* 191 */ "reserved"
};

// 192..255
const char *ncep_level_table[64] = {
/* 192 */ "reserved",
/* 193 */ "reserved",
/* 194 */ "reserved",
/* 195 */ "reserved",
/* 196 */ "reserved",
/* 197 */ "reserved",
/* 198 */ "reserved",
/* 199 */ "reserved",
/* 200 */ "entire atmosphere (considered as a single layer)",
/* 201 */ "entire ocean (considered as a single layer)",
/* 202 */ "reserved",
/* 203 */ "reserved",
/* 204 */ "highest tropospheric freezing level",
/* 205 */ "reserved",
/* 206 */ "grid scale cloud bottom level",
/* 207 */ "grid scale cloud top level",
/* 208 */ "reserved",
/* 209 */ "boundary layer cloud bottom level",
/* 210 */ "boundary layer cloud top level",
/* 211 */ "boundary layer cloud layer",
/* 212 */ "low cloud bottom level",
/* 213 */ "low cloud top level",
/* 214 */ "low cloud layer",
/* 215 */ "cloud ceiling",
/* 216 */ "effective layer top level %gm",
/* 217 */ "effective layer bottom level %gm",
/* 218 */ "effective layer %gm",
/* 219 */ "reserved",
/* 220 */ "planetary boundary layer",
/* 221 */ "layer between two hybrid levels",
/* 222 */ "middle cloud bottom level",
/* 223 */ "middle cloud top level",
/* 224 */ "middle cloud layer",
/* 225 */ "reserved",
/* 226 */ "reserved",
/* 227 */ "reserved",
/* 228 */ "reserved",
/* 229 */ "reserved",
/* 230 */ "reserved",
/* 231 */ "reserved",
/* 232 */ "high cloud bottom level",
/* 233 */ "high cloud top level",
/* 234 */ "high cloud layer",
/* 235 */ "%gC ocean isotherm",
/* 236 */ "layer between two depths below ocean surface",
/* 237 */ "bottom of ocean mixed layer",
/* 238 */ "bottom of ocean isothermal layer",
/* 239 */ "layer ocean surface and 26C ocean isothermal level",
/* 240 */ "ocean mixed layer",
/* 241 */ "%g in sequence",
/* 242 */ "convective cloud bottom level",
/* 243 */ "convective cloud top level",
/* 244 */ "convective cloud layer",
/* 245 */ "lowest level of the wet bulb zero",
/* 246 */ "maximum equivalent potential temperature level",
/* 247 */ "equilibrium level",
/* 248 */ "shallow convective cloud bottom level",
/* 249 */ "shallow convective cloud top level",
/* 250 */ "reserved",
/* 251 */ "deep convective cloud bottom level",
/* 252 */ "deep convective cloud top level",
/* 253 */ "lowest bottom level of supercooled liquid water layer",
/* 254 */ "highest top level of supercooled liquid water layer",
/* 255 */ "missing"
};

// 192..255
const char *kma_level_table[64] = {
/* 192 */ "reserved",
/* 193 */ "reserved",
/* 194 */ "reserved",
/* 195 */ "hybrid sigma pressure",
/* 196 */ "Height from ground KFT",
/* 197 */ "canopy",
/* 198 */ "ICAO convective cloud top level",
/* 199 */ "ICAO convective cloud base level",
/* 200 */ "entire atmosphere (considered as a single layer)",
/* 201 */ "entire ocean (considered as a single layer)",
/* 202 */ "reserved",
/* 203 */ "reserved",
/* 204 */ "highest tropospheric freezing level",
/* 205 */ "reserved",
/* 206 */ "grid scale cloud bottom level",
/* 207 */ "grid scale cloud top level",
/* 208 */ "reserved",
/* 209 */ "boundary layer cloud bottom level",
/* 210 */ "boundary layer cloud top level",
/* 211 */ "boundary layer cloud layer",
/* 212 */ "low cloud bottom level",
/* 213 */ "low cloud top level",
/* 214 */ "low cloud layer",
/* 215 */ "cloud ceiling",
/* 216 */ "reserved",
/* 217 */ "reserved",
/* 218 */ "reserved",
/* 219 */ "reserved",
/* 220 */ "planetary boundary layer",
/* 221 */ "layer between two hybrid levels",
/* 222 */ "middle cloud bottom level",
/* 223 */ "middle cloud top level",
/* 224 */ "middle cloud layer",
/* 225 */ "reserved",
/* 226 */ "reserved",
/* 227 */ "reserved",
/* 228 */ "reserved",
/* 229 */ "reserved",
/* 230 */ "reserved",
/* 231 */ "reserved",
/* 232 */ "high cloud bottom level",
/* 233 */ "high cloud top level",
/* 234 */ "high cloud layer",
/* 235 */ "%gC ocean isotherm",
/* 236 */ "layer between two depths below ocean surface",
/* 237 */ "bottom of ocean mixed layer",
/* 238 */ "bottom of ocean isothermal layer",
/* 239 */ "layer ocean surface and 26C ocean isothermal level",
/* 240 */ "ocean mixed layer",
/* 241 */ "%g in sequence",
/* 242 */ "convective cloud bottom level",
/* 243 */ "convective cloud top level",
/* 244 */ "convective cloud layer",
/* 245 */ "lowest level of the wet bulb zero",
/* 246 */ "maximum equivalent potential temperature level",
/* 247 */ "equilibrium level",
/* 248 */ "shallow convective cloud bottom level",
/* 249 */ "shallow convective cloud top level",
/* 250 */ "reserved",
/* 251 */ "deep convective cloud bottom level",
/* 252 */ "deep convective cloud top level",
/* 253 */ "lowest bottom level of supercooled liquid water layer",
/* 254 */ "highest top level of supercooled liquid water layer",
/* 255 */ "missing"
};

int level1(int mode, int type, int undef_val, float value, int center, int subcenter, char *inv_out);
int level2(int mode, int type1, int undef_val1, float value1, int type2, int undef_val2, 
   float value2, int center, int subcenter, char *inv_out);

int f_lev(ARG0) {

    int level_type1, level_type2;
    float val1, val2;
    int undef_val1, undef_val2;
    int center, subcenter;

    if (mode < 0) return 0;

    center = GB2_Center(sec);
    subcenter = GB2_Subcenter(sec);

    fixed_surfaces(sec, &level_type1, &val1, &undef_val1, &level_type2, &val2, &undef_val2);

    if (mode > 1) {
	if (undef_val1 == 0) sprintf(inv_out,"lvl1=(%d,%lg) ",level_type1,val1);
	else sprintf(inv_out,"lvl1=(%d,missing) ",level_type1);
	inv_out += strlen(inv_out);

	if (undef_val2 == 0) sprintf(inv_out,"lvl2=(%d,%lg):",level_type2,val2);
	else sprintf(inv_out,"lvl2=(%d,missing):",level_type2);
	inv_out += strlen(inv_out);
    }

    level2(mode, level_type1, undef_val1, val1, level_type2, undef_val2, val2, center, subcenter, inv_out);
    return 0;
}

/*
 * level2 is for layers
 */

int level2(int mode, int type1, int undef_val1, float value1, int type2, int undef_val2, float value2, int center, int subcenter,
	char *inv_out) {

    if (type1 == 100 && type2 == 100) {
	sprintf(inv_out,"%g-%g mb",value1/100,value2/100);
    }
    else if (type1 == 102 && type2 == 102) {
	sprintf(inv_out,"%g-%g m above mean sea level",value1,value2);
    }
    else if (type1 == 103 && type2 == 103) {
	sprintf(inv_out,"%g-%g m above ground",value1,value2);
    }
    else if (type1 == 104 && type2 == 104) {
	sprintf(inv_out,"%g-%g sigma layer",value1,value2);
    }
    else if (type1 == 105 && type2 == 105) {
	sprintf(inv_out,"%g-%g hybrid layer",value1,value2);
    }
    else if (type1 == 106 && type2 == 106) {
	/* sprintf(inv_out,"%g-%g m below ground",value1/100,value2/100); removed 1/2007 */
	sprintf(inv_out,"%g-%g m below ground",value1,value2);
    }
    else if (type1 == 107 && type2 == 107) {
	sprintf(inv_out,"%g-%g K isentropic layer",value1,value2);
    }
    else if (type1 == 108 && type2 == 108) {
	sprintf(inv_out,"%g-%g mb above ground",value1/100,value2/100);
    }
    else if (type1 == 111 && type2 == 111) {
	sprintf(inv_out,"%g-%g Eta layer",value1,value2);
    }
    else if (type1 == 115 && type2 == 115) {
	sprintf(inv_out,"%g-%g sigma height layer",value1,value2);
    }
    else if (type1 == 118 && type2 == 118) {
	sprintf(inv_out,"%g-%g hybrid height layer",value1,value2);
    }
    else if (type1 == 119 && type2 == 119) {
	sprintf(inv_out,"%g-%g hybrid pressure layer",value1,value2);
    }
    else if (type1 == 150 && type2 == 150) {
	sprintf(inv_out,"%g-%g generalized vertical height coordinate",value1,value2);
    }
    else if (type1 == 160 && type2 == 160) {
	sprintf(inv_out,"%g-%g m below sea level",value1,value2);
    }
    else if (type1 == 161 && type2 == 161) {
	sprintf(inv_out,"%g-%g m ocean layer",value1,value2);
    }
    else if (type1 == 1 && type2 == 8) {
	sprintf(inv_out,"atmos col");		// compatible with wgrib
    }
    else if (type1 == 9 && type2 == 1) {
	sprintf(inv_out,"ocean column");
    }
    else if (center == NCEP && type1 == 235 && type2 == 235) {
	    sprintf(inv_out,"%g-%gC ocean isotherm layer", value1/10,value2/10);
    }
    else if (center == NCEP && type1 == 236 && type2 == 236) {	// obsolete
	    sprintf(inv_out,"%g-%g m ocean layer", value1*10,value2*10);
    }
    else if (type1 == 255 && type2 == 255) {
	    sprintf(inv_out,"no_level");
    }
    else {
        level1(mode, type1, undef_val1, value1, center, subcenter,inv_out);
	inv_out += strlen(inv_out);
        if (type2 != 255) {
	    sprintf(inv_out," - ");
	    inv_out += strlen(inv_out);
	    level1(mode, type2, undef_val2, value2, center, subcenter,inv_out);
        }
    }
    return 0;
}

/*
 * level1 is for a single level (not a layer)
 */

int level1(int mode, int type, int undef_val, float val, int center, int subcenter,char *inv_out) {

    /* WMO defined levels */

    if (type < 192) {
        if (type == 100 || type == 108) val = val * 0.01;  // Pa -> mb
	sprintf(inv_out,level_table[type], val);
	return 0;
    }

    // no numeric information
    if (type == 255) return 8;

    /* local table for NCEP */
    if (center == NCEP && type >= 192 && type <= 254) {
        if (type == 235) val *= 0.01;  // C -> 0.1C
	sprintf(inv_out,ncep_level_table[type-192], val);
    }

    else if (center == KMA && type >= 192 && type <= 254) {
        if (type == 235) val *= 0.01;  // C -> 0.1C
	sprintf(inv_out,kma_level_table[type-192], val);
    }

    else {
        if (undef_val == 0) sprintf(inv_out,"local level type %d %g", type, val);
        else sprintf(inv_out,"local level type %d", type);
    }

    return 0;
}
